Document pipeline configuration files v2.0

lovromazgon opened the pull request:
This PR changes the pipeline configuration file documentation to explain version 2.0 and adds an explanation for each field. The existing documentation is broken up into 3 separate chapters and slightly expanded.

hariso commented on docs/pipeline-configuration-files/provisioning.mdx:
```diff
@@ -0,0 +1,61 @@
+
+## Pipelines Immutability
+
+Pipelines provisioned by configuration files are **immutable**, any updates needed on a provisioned pipeline have to be
+done through the configuration file. You can still stop or start the pipeline through the UI or API.
```
Nitpick: it feels that "immutable" is not quite technically correct, given that they are mutable, but the changes are restricted to files.

maha-hajja replied:
we could say:
```suggestion
Pipelines provisioned by configuration files are **immutable** through the API, any updates needed on a provisioned pipeline have to be
done through the configuration files. You can only stop or start the pipeline through the UI or API.
```

lovromazgon replied:
Right, I'll try to rephrase it 👍

hariso commented on docs/pipeline-configuration-files/specifications.mdx:
```diff
@@ -0,0 +1,272 @@
+
+- **Data Type**: String
+- **Required**: No
+- **Default**: `builtin:file`
```
Wasn't it the log connector?

lovromazgon replied:
Yes, good catch!

hariso commented on docs/pipeline-configuration-files/specifications.mdx:
```diff
@@ -0,0 +1,272 @@
+- **Data Type**: String
+- **Required**: No
+- **Default**: Same as pipeline [`id`](#id)
+- **Description**: Human readable name for the pipeline. Needs to be unique across all pipelines.
```
Can you please remind me why it has to be unique (when we already have the ID), especially because connector names for example do not need to be unique? I vaguely recall something about our platform and fetching pipelines.

lovromazgon replied:
IIRC it's because we are using the pipeline name in metrics to distinguish them, since names are provided by the user and IDs were not at the time we designed this (they still are not if the pipeline is created through the API). So to prevent mixing metrics between multiple pipelines they need to have unique names.

hariso commented on docs/pipeline-configuration-files/specifications.mdx:
```diff
@@ -0,0 +1,272 @@
+- **Data Type**: String
+- **Required**: Yes
+- **Default**: None
+- **Description**: Defines the processor type.
```
Would be good to also say the available types or where to find the available types.

lovromazgon replied:
I added a link to the processor documentation that we have right now, but we don't have a full list of processors, we just point to the [godoc](https://pkg.go.dev/github.com/conduitio/conduit/pkg/processor/procbuiltin) there. I think we can and should do better by having a full list of processors and their capabilities, similar to the [kafka connect docs](https://docs.confluent.io/platform/current/connect/transforms/overview.html). That's actually what [this issue](https://github.com/ConduitIO/conduit/issues/966) was meant for.

maha-hajja commented on docs/pipeline-configuration-files/specifications.mdx:
````diff
@@ -0,0 +1,272 @@
+      window-nack-threshold: 2  # DLQ nack window threshold
+```
+
+## version
````
love these! this whole file is super detailed and on point.. nice work!

maha-hajja commented on docs/pipeline-configuration-files/specifications.mdx:
```diff
@@ -0,0 +1,272 @@
+- **Required**: Yes
+- **Default**: None
+- **Description**: ID of the [`pipeline`](#pipeline). It should uniquely identify the pipeline across all pipelines. If
+  multiple pipelines use the same ID, none of them will be provisioned.
```
none or the first one?

maha-hajja replied:
oh didn't notice this change! yeah makes sense

maha-hajja commented on docs/pipeline-configuration-files/specifications.mdx:
```diff
@@ -0,0 +1,272 @@
+- **Description**: ID of the [`pipeline`](#pipeline). It should uniquely identify the pipeline across all pipelines. If
+  multiple pipelines use the same ID, none of them will be provisioned.
+
+**Warning**: Changing this property will cause all connectors to start from the beginning.
```
```suggestion
**Warning**: Changing this property will cause all this pipeline's connectors to start from the beginning when Conduit is restarted.
```

maha-hajja commented on docs/pipeline-configuration-files/specifications.mdx:
```diff
@@ -0,0 +1,272 @@
+- **Default**: None
+- **Description**: Human readable description of the pipeline.
+
+### dead-letter-queue
```
I think it's better to have DLQ below the connectors and processors, since they're more important and usually come first

maha-hajja commented on docs/pipeline-configuration-files/specifications.mdx:
```diff
@@ -0,0 +1,272 @@
+- **Description**: ID of the [`connector`](#connector). It should uniquely identify the connector inside the pipeline.
+  If multiple connectors inside the pipeline use the same ID, provisioning will fail.
+
+**Warning**: Changing this property will cause the connector to start from the beginning.
```
```suggestion
**Warning**: Changing this property will cause the connector to start from the beginning when Conduit is restarted.
```

maha-hajja commented on docs/pipeline-configuration-files/specifications.mdx:
```diff
@@ -0,0 +1,272 @@
+- **Required**: Yes
+- **Default**: None
+- **Allowed Values**: `source`, `destination`
+- **Description**: Defines if the connector is a source or a destination. A working pipeline needs at least one source
```
```suggestion
- **Description**: Defines if the connector is a source or a destination. A running pipeline needs at least one source
```

maha-hajja commented on docs/pipeline-configuration-files/specifications.mdx:
```diff
@@ -0,0 +1,272 @@
+- **Data Type**: Mapping node
+- **Required**: Yes
+- **Default**: None
+- **Description**: A map of configuration keys and values for the connector plugin. The values in this map depend on the
```
we could let the user know where to get the needed settings (checking the config table from the connector docs)

simonl2002 commented on docs/pipeline-configuration-files/getting-started.mdx:
````diff
@@ -0,0 +1,73 @@
+          sdk.record.format.options: '{{ printf "%s" .Payload.After }}'
+```
+
+Now start Conduit and notice the log line saying that the pipeline `file-to-file` was created:
````
I'd tweak this slightly. Maybe say something like `Now start Conduit. You should see a log line saying...`

simonl2002 commented on docs/pipeline-configuration-files/provisioning.mdx:
```diff
@@ -0,0 +1,61 @@
+When Conduit starts, it will search for all files ending with `.yml` or `.yaml` in folder `./pipelines` and its
+sub-folders. It will parse the configuration files and provision the pipelines defined in them.
+
+If you want Conduit to search a different directory, or want to provision a single file, you can specify the CLI flag
```
This is implying that you can point `-pipelines.path` to a single file? Is that the case?

lovromazgon replied:
Yup, that's right!

simonl2002 replied:
yeah I tested it myself and came back to just delete this.

simonl2002 commented on docs/pipeline-configuration-files/provisioning.mdx:
```diff
@@ -0,0 +1,61 @@
+invalid configuration value) and won't impact the provisioning of valid configuration files. Conduit will log a warning
+for every error it encounters during this process.
+
+You need to be careful when choosing IDs for your pipelines, connectors and processors. Each pipeline needs to get a
```
Feels like this deserves it's own heading rather than being buried under "ignored errors." It is a pretty important detail.

lovromazgon commented:
Thanks for your reviews! I _think_ I addressed the comments so far, let me know if you see more opportunities to improve the doc 🙂 @maha-hajja @hariso @simonl2002

maha-hajja commented on docs/pipeline-configuration-files/provisioning.mdx:
```diff
@@ -3,59 +3,106 @@ title: 'Provisioning'
+
+Conduit injects environment variables in pipeline configuration files. To
+specify a value using an environment variable, wrap the name of the variable in
+`${}` (e.g. `${MY_ENV_VAR}`). Any value in the configuration can be injected
```
```suggestion
`${}` (e.g. `${MY_ENV_VAR}`). Any value in the configuration file can be injected
```

maha-hajja reviewed (approved):
lgtm

hariso reviewed (approved):
Great work Lovro, especially the specs!
